lib/restbase.js

```diff
@@ -47,10 +47,23 @@ function RESTBase (options, req) {
         this._priv = {
             options: options,
             router: options.router,
-            statsd: options.metrics
+            statsd: options.metrics,
+            reqId: req ? req.headers && req.headers['x-request-id'] || rbUtil.generateRequestId() : null
         };
+    }
 
+    if (req) {
+        req.headers = req.headers || {};
+        if (req.headers['x-request-id']) {
+            // the request has got the request id header set, use that
+            this._priv.reqId = req.headers['x-request-id'];
+        } else if (!this._priv.reqId) {
+            // no request id set, generate it
+            this._priv.reqId = rbUtil.generateRequestId();
+        }
+        req.headers['x-request-id'] = this._priv.reqId;
```

Comment:
I think the main situation where we need to make sure to set the header is for outgoing requests. In server.js we should have access to the child restbase instance. We could add an accessor (getRequestId or the like) that returns `this._priv.reqId` and use that for logging from server.js.